File: tools_box/logistics/report/vehicle_schedule_report/vehicle_schedule_report.py
# Copyright (c) 2013, masonarmani38@gmail.com and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _


def execute(filters=None):
    cconditions = conditions = "(1=1)"

    if filters.get('from') and filters.get('to'):
        conditions += " and date BETWEEN DATE('{from}') and DATE('{to}')"

    if filters.get('vehicle'):
        conditions += " and vehicle = '{vehicle}'"

    if filters.get('type'):
        conditions += " and type = '{type}'"

    if filters.get('ref_name'):
        cconditions += " and ref_name = '{ref_name}'"

    # get all parents
    parents = frappe.db.sql(
        "SELECT name, '' customer, date,vehicle, type, daily_cost ,'' ref_name, total_amount amount  from `tabVehicle Schedule` "
        "where {0}".format(conditions.format(**filters)), as_dict=1)
    data = []
    for index, parent in enumerate(parents):
        rat = (parent.daily_cost / parent.amount) * 100
        parent.update({"parent": None, "indent": 0, "has_value": True, "ratio": rat})
        data.append(parent)

        if filters.get('type') != "Operations":
            children = frappe.db.sql(
                "SELECT name, ref_name , party, amount from `tabVehicle Schedule {2} Item` where {0} and parent = '{1}' "
                    .format(cconditions.format(**filters), parent.name, filters.get('type')), as_dict=1)

            for child in children:
                data.append(
                    dict(name="", customer=child.get('party'), date="", vehicle="",
                         type="", daily_cost="", ref_name=child.get('ref_name'),
                         indent=1, amount=child.get('amount'), ratio=0,
                         parent=parent.get('name'), has_value=True)
                )

    doc = "Delivery Note"
    if filters.get('type') == "Inbound":
        doc = "Purchase Order"

    return get_columns(doc), data


def get_columns(doc):
    return [{
        "fieldname": "name",
        "label": _("ID"),
        "fieldtype": "Link",
        "options": "Vehicle Schedule",
        "width": 160
    }, {
        "fieldname": "customer",
        "label": _("Customer"),
        "fieldtype": "data",
        "width": 160
    }, {
        "fieldname": "date",
        "label": _("Date"),
        "fieldtype": "Date",
        "width": 120
    }, {
        "fieldname": "vehicle",
        "label": _("Vehicle"),
        "fieldtype": "Link",
        "options": "Vehicle",
        "width": 120
    }, {
        "fieldname": "type",
        "label": _("Type"),
        "fieldtype": "Data",
        "options": "data",
        "width": 80
    }, {
        "fieldname": "daily_cost",
        "label": _("Daily Cost"),
        "fieldtype": "Currency",
        "options": "currency",
        "width": 120
    }, {
        "fieldname": "ref_name",
        "label": _(doc),
        "fieldtype": "Link",
        "options": doc,
        "width": 120
    }, {
        "fieldname": "amount",
        "label": _("Amount"),
        "fieldtype": "Currency",
        "options": "currency",
        "width": 140
    }, {
        "fieldname": "ratio",
        "label": _("Ratio"),
        "fieldtype": "Float",
        "options": "",
        "width": 100
    }]

File: tools_box/logistics/report/vehicle_schedule_report/vehicle_schedule_report.js
// Copyright (c) 2016, masonarmani38@gmail.com and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Vehicle Schedule Report"] = {
    "filters": [
        {
            "fieldname": "from",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default":"2018-01-01",
            "width": "80",
            "reqd": 1
        },
        {
            "fieldname": "to",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        },
        {
            "fieldname": "vehicle",
            "label": __("Vehicle"),
            "fieldtype": "Link",
            "options": "Vehicle",
            "reqd": 0
        },
        {
            "fieldname": "type",
            "label": __("Type"),
            "fieldtype": "Select",
            "reqd": 0,
            "default": "Outbound",
            "options": [
                "Inbound", "Outbound", "Operations"
            ],
        },
        {
            "fieldname": "ref_name",
            "label": __("Reference Document"),
            "fieldtype": "Link",
            "options": "Purchase Order",
            "get_query": function () {
                //
                var type = frappe.query_report.filters[3].value;
                return {
                    "doctype": (type == "Inbound") ? "Purchase Order" : "Delivery Note",
                    "filters": {
                        docstatus: 1
                    }
                }
            }
        }
    ],
    "formatter": function (row, cell, value, columnDef, dataContext, default_formatter) {
        // formatter is called on every cell and the formatted values is then returned
        // value contains the first row value from the server like item[index]
        // cell contains the index the current cell definition
        // columnDef contains information about the current cell
        // default_formatter is the default formatter
        if (true){
            const _context = dataContext;
            if (columnDef.df.fieldname == "name") {
                value = _context[columnDef.df.fieldname];
                //columnDef.df.link_onclick =
                    //"frappe.query_reports['Vehicle Schedule Report'].low(" + JSON.stringify(dataContext) + ")";
                columnDef.df.is_tree = true;
            }

            value = default_formatter(row, cell, value, columnDef, dataContext);

            if (!_context.parent && columnDef.df.fieldname == "name") {
                var $value = $(value).css("font-weight", "bold");
                console.log($(value).html())
                value = $value.html()
                console.log(value)
            }
        }


        return value;
    },
    "tree": true,
    "name_field": "name",
    "parent_field": "parent",
    "initial_depth": 3,
    "onload":function(report){
        //console.log(report)
    }
}
